lib/spack/spack/cmd/fetch.py

```diff
@@ -27,13 +28,24 @@ def setup_parser(subparser):
 
 
 def fetch(parser, args):
-    if not args.specs:
-        tty.die("fetch requires at least one package argument")
+    if args.specs:
+        specs = spack.cmd.parse_specs(args.specs, concretize=True)
+    else:
+        # No specs were given explicitly, check if we are in an
+        # environment. If yes fetch all uninstalled specs from
+        # it.
+        env = ev.get_env(args, 'fetch')
+        if env:
+            specs = env.uninstalled_specs()
```

Comment:
I think we should fetch all packages in the environment, not just uninstalled specs. If they are already present in the cache, they will be skipped automatically.

Reply:
Are you sure? Fetching provides the sources, which may no longer be around once things are installed. And we do not need them as we won't be building those anymore. I don't really see an upside to just fetching all

Reply:
I think the usual reason someone will try to fetch an entire environment (at least at our site) is to get all the sources so that they can build it somewhere else.  They may have installed some things *right here* but they probably want to build those someplace else too.

Can you make this fetch everything by default?  IMO, the end state of the operation should by default be idempotent -- I run `spack fetch` in an environment, and I have everything I need to build it.

Fetching uninstalled can be an option as I can see how that's useful but I do think it's important that you can say "I ran spack fetch and therefore I could rebuild all of this".